### PR TITLE
Improved error handling around sending data in Socket/SocketStream (fixes #15614)

### LINKED ISSUE #15614: Socket closed on the server side, indefinitely allows writing on the client side
**Bug description**
Socket closed on the server side, indefinitely allows writing on the client side

**To Reproduce**
Evalute:
```
	| listener clientSocket serverSocket clientStream serverStream writeResult connectedResult |
	listener := Socket newTCP.
	[listener listenOn: 0 backlogSize: 4.

	clientSocket := Socket newTCP.
	clientSocket connectTo: #[127 0 0 1] port: listener localPort.
	clientSocket waitForConnectionFor: 1.
	self assert: clientSocket isConnected.

	serverSocket := listener waitForAcceptFor: 1.
	self assert: serverSocket isConnected.
	] ensure:[listener destroy].

	clientStream := SocketStream on: clientSocket.
	serverStream := SocketStream on: serverSocket.
	
	
	serverStream close.
	writeResult  := [clientStream nextPutAll:'A line of text'; flush.
	true] on: ConnectionClosed do:[:ex | false].  
	connectedResult := clientStream isConnected.
	
	clientStream ifNotNil:[clientStream destroy].
	serverStream ifNotNil:[serverStream destroy].

	writeResult -> connectedResult
```

**Expected behavior**
The evaluation above should result in false -> false meaning that:
a write operation on a disconnected socket should fail, possibly with a ConnectionClosed errors or as a bare minimum #isConnected should return false after the first write operation failure.

**Version information:**
 - OS: at least on Windows and Linux
 - Version: [e.g. 1.6.1]
 - Pharo Version 10

**Expected development cost**
A few hours for somebody who knows the area.
I would says it's more concerning the impact for existing code.

### PR DESCRIPTION
**Pull request by daniels220**
Distinguish timeout and connection-closed, and actually pass errors in SocketStream>>flush rather than always swallowing them.

I believe this addresses the core issues in #15614. Adding a `SocketStream>>waitForSendDone` or similar could be a separate issue if desired.

### CODE REVIEW

File: src/Network-Kernel/Socket.class.st [hunk 1]
```diff
@@ -1639,18 +1630,14 @@ Socket >> waitForDataFor: timeout [
...
-
-	self dataAvailable
-		ifFalse: [ ^ (self isConnected
-				ifTrue: [ timedOutBlock ]
-				ifFalse: [ closedBlock ]) value ]
+	[ self dataAvailable ] whileFalse: [
+		self isConnected ifFalse: [ ^ closedBlock value ].
+		(msecsElapsed := Time millisecondsSince: startTime) < msecsDelta
+			ifFalse: [ ^ timedOutBlock value ].
+		self readSemaphore waitTimeoutMilliseconds:: msecsDelta - msecsElapsed ]
```
Comment (tesonep) [inline src/Network-Kernel/Socket.class.st]: Here there is a typo (two colons :: ) and it should be fixed so we can integrate it.

**Review by tesonep — CHANGES_REQUESTED** [review]
Thanks for the contribution, please fix the typo so we can integrate it

File: src/Network-Kernel/Socket.class.st (hunk 1 above)
Comment (daniels220) [reply src/Network-Kernel/Socket.class.st]: D'oh! Bad cherry-pick—I made these changes in an image where that method was still `waitTimeoutMSecs:` and Git did an _unusually_ bad job of detecting _what_ the conflict actually was so I had to do a lot of manual editing. Fixed now.

**Comment by daniels220** [comment]
@tesonep Anything stopping this from being merged now that the typo is fixed? I realize it's effectively a prerequisite to #16073—merging this may actually cause a merge conflict, which I'll be happy to resolve, but they're closely related.

**Comment by guillep** [comment]
Just we are busy, thanks for the heads up ^^

**Comment by daniels220** [comment]
I understand—I've had a couple PRs I knew I wanted to _make_ for a month or two before I found the time. Thanks for merging. Sure enough there was a conflict with #16073, resolved now.

**Pull request merged**